File: v2/web/lib/runtime-state.ts
import { promises as fs } from "fs";
import path from "path";

export interface WriteFreezeState {
  writeFrozen: boolean;
  reason: string;
  message: string;
  source: string;
  createdAt: string;
}

export interface CutoverState {
  phase: string;
  status: string;
  rollbackAuthority: string;
  cutbackRequired: boolean;
  reason: string;
  message: string;
  source: string;
  createdAt: string;
  manifestPath: string;
}

function markerPath(): string {
  return process.env.FUNNEL_V2_WRITE_FREEZE_MARKER || path.resolve(process.cwd(), "../contracts/hosted-runtime/write-freeze.json");
}

function cutoverMarkerPath(): string {
  return process.env.FUNNEL_V2_CUTOVER_STATE_PATH || path.resolve(process.cwd(), "../contracts/hosted-runtime/cutover-state.json");
}

export async function readWriteFreezeState(): Promise<WriteFreezeState> {
  const defaultState: WriteFreezeState = {
    writeFrozen: false,
    reason: "",
    message: "",
    source: "",
    createdAt: "",
  };
  try {
    const raw = await fs.readFile(markerPath(), "utf-8");
    const parsed = JSON.parse(raw) as Record<string, unknown>;
    return {
      ...defaultState,
      writeFrozen: Boolean(parsed.write_frozen ?? true),
      reason: String(parsed.reason || ""),
      message: String(parsed.message || ""),
      source: String(parsed.source || ""),
      createdAt: String(parsed.created_at || ""),
    };
  } catch {
    return defaultState;
  }
}

export async function readCutoverState(): Promise<CutoverState> {
  const defaultState: CutoverState = {
    phase: "idle",
    status: "idle",
    rollbackAuthority: "sqlite",
    cutbackRequired: false,
    reason: "",
    message: "",
    source: "",
    createdAt: "",
    manifestPath: "",
  };
  try {
    const raw = await fs.readFile(cutoverMarkerPath(), "utf-8");
    const parsed = JSON.parse(raw) as Record<string, unknown>;
    return {
      ...defaultState,
      phase: String(parsed.phase || "idle"),
      status: String(parsed.status || "idle"),
      rollbackAuthority: String(parsed.rollback_authority || "sqlite"),
      cutbackRequired: Boolean(parsed.cutback_required || false),
      reason: String(parsed.reason || ""),
      message: String(parsed.message || ""),
      source: String(parsed.source || ""),
      createdAt: String(parsed.created_at || ""),
      manifestPath: String(parsed.manifest_path || ""),
    };
  } catch {
    return defaultState;
  }
}

File: v2/api/app/runtime_state.py
from __future__ import annotations

import json
from pathlib import Path
from typing import Any

from app.config import Settings


def default_hosted_runtime_dir(settings: Settings) -> Path:
    return settings.contract_dir / "hosted-runtime"


def write_freeze_marker_path(settings: Settings) -> Path:
    return settings.write_freeze_marker_path


def cutover_state_path(settings: Settings) -> Path:
    return settings.cutover_state_path


def read_write_freeze_state(settings: Settings) -> dict[str, Any]:
    marker_path = write_freeze_marker_path(settings)
    default_payload: dict[str, Any] = {
        "write_frozen": False,
        "reason": "",
        "message": "",
        "source": "",
        "created_at": "",
    }
    if not marker_path.exists():
        return default_payload
    try:
        payload = json.loads(marker_path.read_text(encoding="utf-8"))
    except (OSError, json.JSONDecodeError):
        return {
            **default_payload,
            "write_frozen": True,
            "reason": "invalid_marker",
            "message": "Writes are temporarily frozen while hosted maintenance state is being reconciled.",
            "source": str(marker_path),
        }
    return {
        **default_payload,
        **payload,
        "write_frozen": bool(payload.get("write_frozen", True)),
    }


def write_frozen(settings: Settings) -> bool:
    return bool(read_write_freeze_state(settings).get("write_frozen"))


def read_cutover_state(settings: Settings) -> dict[str, Any]:
    marker_path = cutover_state_path(settings)
    default_payload: dict[str, Any] = {
        "phase": "idle",
        "status": "idle",
        "rollback_authority": "sqlite",
        "cutback_required": False,
        "reason": "",
        "message": "",
        "source": "",
        "created_at": "",
        "manifest_path": "",
    }
    if not marker_path.exists():
        return default_payload
    try:
        payload = json.loads(marker_path.read_text(encoding="utf-8"))
    except (OSError, json.JSONDecodeError):
        return {
            **default_payload,
            "phase": "unknown",
            "status": "invalid_marker",
            "cutback_required": True,
            "reason": "invalid_marker",
            "message": "Cutover state marker is unreadable. Treat the hosted stack as rollback-only until it is reconciled.",
            "source": str(marker_path),
        }
    return {
        **default_payload,
        **payload,
        "cutback_required": bool(payload.get("cutback_required", False)),
    }


def set_write_freeze(
    settings: Settings,
    *,
    enabled: bool,
    reason: str = "",
    message: str = "",
    source: str = "",
    created_at: str = "",
) -> dict[str, Any]:
    marker_path = write_freeze_marker_path(settings)
    marker_path.parent.mkdir(parents=True, exist_ok=True)
    if not enabled:
        if marker_path.exists():
            marker_path.unlink()
        return read_write_freeze_state(settings)
    payload = {
        "write_frozen": True,
        "reason": reason,
        "message": message or "Writes are temporarily frozen while hosted maintenance runs.",
        "source": source,
        "created_at": created_at,
    }
    marker_path.write_text(json.dumps(payload, indent=2), encoding="utf-8")
    return payload


def set_cutover_state(
    settings: Settings,
    *,
    phase: str,
    status: str,
    rollback_authority: str = "sqlite",
    cutback_required: bool = False,
    reason: str = "",
    message: str = "",
    source: str = "",
    created_at: str = "",
    manifest_path: str = "",
) -> dict[str, Any]:
    marker_path = cutover_state_path(settings)
    marker_path.parent.mkdir(parents=True, exist_ok=True)
    payload = {
        "phase": phase,
        "status": status,
        "rollback_authority": rollback_authority,
        "cutback_required": bool(cutback_required),
        "reason": reason,
        "message": message,
        "source": source,
        "created_at": created_at,
        "manifest_path": manifest_path,
    }
    marker_path.write_text(json.dumps(payload, indent=2), encoding="utf-8")
    return payload
